core/src/main/java/com/netflix/iceberg/BaseTransaction.java

```diff
@@ -0,0 +1,281 @@
+  static Transaction createTableTransaction(TableOperations ops, TableMetadata start) {
+    Preconditions.checkArgument(ops.current() == null,
+        "Cannot start create table transaction: table already exists");
+    return new BaseTransaction(ops, start);
+  }
+
+  static Transaction newTransaction(TableOperations ops) {
+    return new BaseTransaction(ops, ops.refresh());
+  }
+
+  private final TransactionTable transactionTable;
+  private final TableOperations ops;
+  private final TableOperations transactionOps;
+  private final List<PendingUpdate> updates;
+  private TableMetadata base;
+  private TableMetadata lastBase;
+  private TableMetadata current;
+
+  private BaseTransaction(TableOperations ops, TableMetadata start) {
+    this.transactionTable = new TransactionTable();
+    this.ops = ops;
+    this.transactionOps = new TransactionTableOperations();
+    this.updates = Lists.newArrayList();
+    this.base = ops.current();
+    this.lastBase = null;
+    this.current = start;
+  }
+
+  @Override
+  public Table table() {
+    return transactionTable;
+  }
+
+  private void checkLastOperationCommitted(String operation) {
+    Preconditions.checkState(lastBase != current,
+        "Cannot create new %s: last operation (%s) is not committed", operation,
+        updates.get(updates.size() - 1).getClass().getSimpleName());
+    this.lastBase = current;
+  }
+
+  @Override
+  public UpdateProperties updateProperties() {
+    checkLastOperationCommitted("UpdateProperties");
+    UpdateProperties props = new PropertiesUpdate(transactionOps);
+    updates.add(props);
+    return props;
+  }
+
+  @Override
+  public AppendFiles newAppend() {
+    checkLastOperationCommitted("AppendFiles");
+    AppendFiles append = new MergeAppend(transactionOps);
+    updates.add(append);
+    return append;
+  }
+
+  @Override
+  public RewriteFiles newRewrite() {
+    checkLastOperationCommitted("RewriteFiles");
+    RewriteFiles rewrite = new ReplaceFiles(transactionOps);
+    updates.add(rewrite);
+    return rewrite;
+  }
+
+  @Override
+  public DeleteFiles newDelete() {
+    checkLastOperationCommitted("DeleteFiles");
+    DeleteFiles delete = new StreamingDelete(transactionOps);
+    updates.add(delete);
+    return delete;
+  }
+
+  @Override
+  public ExpireSnapshots expireSnapshots() {
+    checkLastOperationCommitted("ExpireSnapshots");
+    ExpireSnapshots expire = new RemoveSnapshots(transactionOps);
+    updates.add(expire);
+    return expire;
+  }
+
+  @Override
+  public void commitTransaction() {
+    if (base != null) {
+      // first try to commit the update chain as-is in case it will succeed
+      boolean committed = false;
+      if (base == ops.refresh()) {
+        try {
+          ops.commit(base, current);
+          committed = true;
+        } catch (CommitFailedException e) {
+          // will retry because committed is false
+        }
+      }
+
+      if (!committed) {
```

Comment:
I feel like we might want to include transaction properties to control behavior of a transaction.  We're assuming a transaction follows the same optimistic approach as an atomic operation on the table, but this may not be the case for things we consider "transactional".  For example, there could be cases were we expect that there will no intervening commits and the transaction commit should fail if there are.

Reply:
That would take care of this particular case, but that was just an example.  Other systems have a number of transaction behaviors that are controllable (auto commit, etc.).  I'm not sure how applicable these are to current use cases.  I think this is fine for now, but we might want to formalize some of the commit behaviors and expose them (like read consistency, etc.).

Reply:
I agree, but I think this should be okay for now. We'll need to carefully consider what types of guarantees we make. This is pretty restrictive, so I think it will be okay.